R/teal_slice.R

```diff
@@ -1,301 +1,198 @@
+      "fixed", "locked", "multiple"
     )
+    formal_args <- formal_args[ts_var_args]
+    args <- c(formal_args, list(...))
+    checkmate::assert_string(varname)
+    checkmate::assert_multi_class(choices, .filterable_class, null.ok = TRUE)
+    checkmate::assert_multi_class(selected, .filterable_class, null.ok = TRUE)
+    checkmate::assert_flag(keep_na, null.ok = TRUE)
+    checkmate::assert_flag(keep_inf, null.ok = TRUE)
+    checkmate::assert_flag(multiple)
+    if (missing(id)) {
+      args$id <- paste(
+        Filter(length, args[c("dataname", "varname", "experiment", "arg")]),
+        collapse = " "
+      )
+    } else {
+      checkmate::assert_string(id)
+    }
+    ans <- do.call(shiny::reactiveValues, args)
+    class(ans) <- c("teal_slice_var", "teal_slice", class(ans))
+    ans
+  } else {
+    stop("Must provide either `expr` or `varname`.")
   }
-  checkmate::assert_list(exclude_varnames, names = "named", types = "character", null.ok = TRUE, min.len = 1)
-  checkmate::assert_list(include_varnames, names = "named", types = "character", null.ok = TRUE, min.len = 1)
-  checkmate::assert_character(count_type, len = 1, null.ok = TRUE)
-  checkmate::assert_subset(count_type, choices = c("all", "none"), empty.ok = TRUE)
-  checkmate::assert_logical(module_add)
-
-  structure(
-    slices,
-    exclude_varnames = exclude_varnames,
-    include_varnames = include_varnames,
-    count_type = count_type,
-    module_add = module_add,
-    class = c("teal_slices", class(slices))
-  )
 }
 
-
-# teal_slice methods ----
-
-# check for teal_slice
 #' @rdname teal_slice
+#' @export
 #' @keywords internal
 #'
 is.teal_slice <- function(x) { # nolint
   inherits(x, "teal_slice")
 }
 
-
-# convert list to teal_slice
 #' @rdname teal_slice
+#' @export
 #' @keywords internal
 #'
 as.teal_slice <- function(x) { # nolint
   checkmate::assert_list(x, names = "named")
-
-  fun <- if ("expr" %in% names(x)) {
-    filter_expr
-  } else {
-    filter_var
-  }
-  do.call(fun, x)
+  do.call(teal_slice, x)
 }
 
-# concatenate method for teal_slice
-#' @export
 #' @rdname teal_slice
+#' @export
 #' @keywords internal
 #'
-c.teal_slice <- function(...) {
-  ans <- unlist(list(...), recursive = FALSE)
-  if (anyDuplicated(names(ans))) {
-    ans <- ans[!duplicated(names(ans))]
-    warning("duplicate field names were discarded")
-  }
-  class(ans) <- c("teal_slice", class(ans))
-  ans
-}
-
-#' @keywords internal
-#' @export
 as.list.teal_slice <- function(x, ...) {
-  formals <- if (inherits(x, "teal_slice_expr")) {
-    formals(filter_expr)
-  } else {
-    formals(filter_var)
-  }
+  formal_args <- setdiff(names(formals(teal_slice)), "...")
 
   x <- if (shiny::isRunning()) {
     shiny::reactiveValuesToList(x)
   } else {
     shiny::isolate(shiny::reactiveValuesToList(x))
   }
 
-  formal_args <- setdiff(names(formals), "...")
+  formal_args <- intersect(formal_args, names(x))
   extra_args <- setdiff(names(x), formal_args)
```

Comment:
Where exactly? I don't get the idea

Reply:
`extra_args <- rev(setdiff(names(x), formal_args))`

Se, we have seen that `reactiveValuesToList` returns a list where the elements are in a different order to the one in the original `reactiveValues`. On Mac the order seems unpredictable but on Linux it is the reverse of the original.
By element names to names of formal arguments of `filter_var` we reorder some of the elements (here, `formal_args`), but the others (`extra_args`) remain mixed up. In case they are merely reversed, a `rev` will restore them to the original order.